File: src/catalog/sync.ts
import fs from 'node:fs';
import path from 'node:path';
import { config } from '../config.js';
import { logger } from '../logger.js';
import type { IncrementalMode } from '../types.js';
import { AppDb } from '../storage/db.js';
import { ElcomClient } from './elcomClient.js';

export class CatalogSyncService {
  constructor(
    private readonly db: AppDb,
    private readonly client: ElcomClient,
  ) {}

  async initialSync(): Promise<{ products: number }> {
    const products = await this.client.getProductsScrollAll();
    this.db.upsertProducts(products);
    this.db.setMetadata('catalog.last_initial_sync', new Date().toISOString());
    await this.refreshFullTreeIfNeeded(true);

    logger.info({ products: products.length }, 'Initial sync completed');
    return { products: products.length };
  }

  async incrementalSync(mode: IncrementalMode): Promise<{ products: number }> {
    const products = await this.client.getProductsIncremental(mode);
    if (products.length > 0) {
      this.db.upsertProducts(products);
    }

    this.db.setMetadata(`catalog.last_incremental_sync.${mode}`, new Date().toISOString());
    await this.refreshFullTreeIfNeeded(false);

    logger.info({ products: products.length, mode }, 'Incremental sync completed');
    return { products: products.length };
  }

  private async refreshFullTreeIfNeeded(force: boolean): Promise<void> {
    const key = 'catalog.last_full_tree_sync';
    const last = this.db.getMetadata(key);
    const monthMs = 1000 * 60 * 60 * 24 * 30;

    if (!force && last) {
      const delta = Date.now() - new Date(last).getTime();
      if (delta < monthMs) {
        return;
      }
    }

    const tree = await this.client.getCatalogFullTree();
    const treePath = path.join(config.outputDir, 'catalog-full-tree.json');
    fs.mkdirSync(path.dirname(treePath), { recursive: true });
    fs.writeFileSync(treePath, JSON.stringify(tree, null, 2), 'utf8');
    this.db.setMetadata(key, new Date().toISOString());

    logger.info({ treePath }, 'Catalog full-tree refreshed');
  }
}

File: internal/catalog/sync.go
package catalog

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"time"

	"elcom/internal/config"
	"elcom/internal/storage"
)

type SyncService struct {
	db     *storage.DB
	client *Client
	cfg    config.Config
}

func NewSyncService(db *storage.DB, cfg config.Config) *SyncService {
	return &SyncService{db: db, client: NewClient(cfg), cfg: cfg}
}

func (s *SyncService) InitialSync(ctx context.Context) (int, error) {
	products, err := s.client.GetProductsScrollAll(ctx)
	if err != nil {
		return 0, err
	}
	if err := s.db.UpsertProducts(products); err != nil {
		return 0, err
	}
	_ = s.db.SetMetadata("catalog.last_initial_sync", time.Now().UTC().Format(time.RFC3339))
	if err := s.refreshFullTreeIfNeeded(ctx, true); err != nil {
		return 0, err
	}
	return len(products), nil
}

func (s *SyncService) IncrementalSync(ctx context.Context, mode string) (int, error) {
	products, err := s.client.GetProductsIncremental(ctx, mode)
	if err != nil {
		return 0, err
	}
	if len(products) > 0 {
		if err := s.db.UpsertProducts(products); err != nil {
			return 0, err
		}
	}
	_ = s.db.SetMetadata("catalog.last_incremental_sync."+mode, time.Now().UTC().Format(time.RFC3339))
	if err := s.refreshFullTreeIfNeeded(ctx, false); err != nil {
		return 0, err
	}
	return len(products), nil
}

func (s *SyncService) refreshFullTreeIfNeeded(ctx context.Context, force bool) error {
	const key = "catalog.last_full_tree_sync"
	last, err := s.db.GetMetadata(key)
	if err != nil {
		return err
	}

	if !force && last != nil {
		if parsed, err := time.Parse(time.RFC3339, *last); err == nil {
			if time.Since(parsed) < 30*24*time.Hour {
				return nil
			}
		}
	}

	tree, err := s.client.GetCatalogFullTree(ctx)
	if err != nil {
		return err
	}
	blob, _ := json.MarshalIndent(tree, "", "  ")
	treePath := filepath.Join(s.cfg.OutputDir, "catalog-full-tree.json")
	if err := os.MkdirAll(filepath.Dir(treePath), 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(treePath, blob, 0o644); err != nil {
		return err
	}
	return s.db.SetMetadata(key, time.Now().UTC().Format(time.RFC3339))
}
